relayer/src/service.ts

```diff
@@ -0,0 +1,126 @@
+import { CrossChainMessenger, MessageStatus } from "@eth-optimism/sdk";
+import { Provider } from "@ethersproject/abstract-provider";
+
+type MessageRelayerOptions = {
+  l1RpcProvider: Provider
+  l2RpcProvider: Provider
+  l1Wallet: Signer
+  fromL2TransactionHash: string
+}
+
+type MessageRelayerState = {
+  wallet: Signer
+  messenger: CrossChainMessenger
+  L2TxBlockNumber: number
+}
+
+export class MessageRelayerService extends BaseServiceV2<
+  MessageRelayerOptions,
+  Record<string, never>,
+  MessageRelayerState
+> {
+  constructor(options?: Partial<MessageRelayerOptions>) {
+    super({
+      name: "Message_Relayer",
+      options,
+      optionsSpec: {
+        l1RpcProvider: {
+          validator: validators.provider,
+          desc: "Provider for interacting with L1.",
+        },
+        l2RpcProvider: {
+          validator: validators.provider,
+          desc: "Provider for interacting with L2.",
+        },
+        l1Wallet: {
+          validator: validators.wallet,
+          desc: "Wallet used to interact with L1.",
+        },
+        fromL2TransactionHash: {
+          validator: validators.str,
+          desc: "Index of the first L2 transaction to start processing from.",
+        },
+      },
+      metricsSpec: { },
+    });
+  }
+
+  protected async init(): Promise<void> {
+    this.state.wallet = this.options.l1Wallet.connect(
+      this.options.l1RpcProvider
+    );
+
+    const l1Network = await this.state.wallet.provider.getNetwork();
+    const l1ChainId = l1Network.chainId;
+    this.state.messenger = new CrossChainMessenger({
+      l1SignerOrProvider: this.state.wallet,
+      l2SignerOrProvider: this.options.l2RpcProvider,
+      l1ChainId,
+    });
+
+    const receipt = await this.state.messenger.l2Provider.getTransactionReceipt(this.options.fromL2TransactionHash);
+    this.state.L2TxBlockNumber = receipt.blockNumber;
+  }
+
+  protected async main(): Promise<void> {
+    // Get messages from transaction
+    const messages = await this.state.messenger.getMessagesByTransaction(
+      this.options.fromL2TransactionHash
+    );
+
+    // No messages in this transaction so we can move on to the next one.
+    if (messages.length === 0) {
+      this.logger.warn(`no message found in L2 transaction ${this.options.fromL2TransactionHash}`);
+      return;
+    }
+
+    // Make sure that all messages sent within the transaction are finalized. If any messages
+    // are not finalized, then we're going to break the loop which will trigger the sleep and
+    // wait for a few seconds before we check again to see if this transaction is finalized.
+    let isFinalized = true;
+    for (const message of messages) {
+      const status = await this.state.messenger.getMessageStatus(message);
+      if (
+        status === MessageStatus.IN_CHALLENGE_PERIOD ||
+        status === MessageStatus.STATE_ROOT_NOT_PUBLISHED
+      ) {
+        isFinalized = false;
```

Comment:
It might, yes. Here we need to decide if the relayer should block until this is possible - which will make process handling more complex.

Alternatively we could just throw an error here and let the agent respawn the process in this case.

Reply:
I'm leaning more towards just making this a short-lived process.

Reply:
Yes, I agree with you. 👍🏾